src/CodeGen_LLVM.cpp

```diff
@@ -2817,156 +2819,35 @@ void CodeGen_LLVM::visit(const Call *op) {
+        // It is assumed that the slot index is valid for the given typed struct.
         //
-        // Note that both the instance and definition are needed because the instance
-        // is often a void* by the time it is -- it will have been been passed through
-        // an API that takes an opaque pointer as a void*, and needs explicit casting
-        // back to the correct type for safe field access.
-        //
-        // It is assumed that the slot index is valid for the given typed_struct.
-        //
-        // TODO: this comment is replicated in CodeGen_C and should be updated there too.
+        // TODO: this comment is replicated in CodeGen_LLVM and should be updated there too.
         // TODO: https://github.com/halide/Halide/issues/6468
-
-        // TODO(zalman): Validate the Halide type of the result matches the LLVM type?
-        // TODO(zalman): Are struct types the right level of indirection.
         internal_assert(op->args.size() == 3);
-        llvm::Value *untyped_struct_instance = codegen(op->args[0]);
-        llvm::Value *typed_struct_definition = codegen(op->args[1]);
-        const uint64_t *index = as_const_uint(op->args[2]);
-        internal_assert(index != nullptr);
-        llvm::Type *struct_type = typed_struct_definition->getType();
-        llvm::Value *typed_struct_instance = builder->CreatePointerCast(untyped_struct_instance, struct_type);
-        llvm::Value *gep = CreateInBoundsGEP(builder, typed_struct_instance, {ConstantInt::get(i32_t, 0), ConstantInt::get(i32_t, (int)*index)});
-        value = builder->CreateLoad(gep->getType()->getPointerElementType(), gep);
-    } else if (op->is_intrinsic(Call::resolve_function_name)) {
-        internal_assert(op->args.size() == 1);
-        const Call *decl_call = op->args[0].as<Call>();
-        internal_assert(decl_call != nullptr);
-        std::string name;
-        if (decl_call->call_type == Call::ExternCPlusPlus) {
-            user_assert(get_target().has_feature(Target::CPlusPlusMangling)) << "Target must specify C++ name mangling (\"c_plus_plus_name_mangling\") in order to call C++ externs. (" << decl_call->name << ")\n";
-
-            std::vector<std::string> namespaces;
-            name = extract_namespaces(decl_call->name, namespaces);
-            std::vector<ExternFuncArgument> mangle_args;
-            for (const auto &arg : decl_call->args) {
-                mangle_args.emplace_back(arg);
-            }
-            name = cplusplus_function_mangled_name(name, namespaces, decl_call->type, mangle_args, get_target());
+        llvm::Value *struct_instance = codegen(op->args[0]);
+        llvm::Value *struct_prototype = codegen(op->args[1]);
+        llvm::Value *typed_struct_instance = builder->CreatePointerCast(struct_instance, struct_prototype->getType());
+        const int64_t *index = as_const_int(op->args[2]);
+
+        // make_struct can use a fixed-size struct, an array type, or a scalar
```

Comment:
Not to sidetrack, but: Wait, seriously? Is that pre-existing behavior? If so, why? Presumably there's a benefit that is worth the complication?

Reply:
Yes, it cleans up the LLVM IR and generally makes more optimizations happen, especially when you have a lot of things of the same type in a make_struct call (e.g. due to tracing, where it's all ints)